### PR TITLE
ESP32 spi slave mode

### LINKED ISSUE #469: SPI slave support
(first of all, thanks for all the amazing progress with Rust support for the ESP32, it truly is amazing!)

I'm trying to build an ISP slave device using the ESP32, and of course I want to build it using Rust. However, it _seems_ like there is no support fir this currently?

If I understand correctly, the SPI traits in embedded-hal is only for SPI master, and when slave traits will be added they will be separate and probably named spi_slave. ref: https://github.com/rust-embedded/embedded-hal/pull/396. And as far as I understand, esp-hal is an implementation of embedded-hal?

It seems like the only way right now is to use Rust with the ESP IDF (e.g. using ), and then call the C functions (`spi_slave_initialize`, `spi_slave_transmit`, etc.) via the `esp_idf_sys` crate?

If my understanding is correct, than this is a feature request to add SPI slave support. In either case, any pointers or guidance would be much appreciated!

---

- [ ] ESP32 (Mode 1 and 3) #2278
- [x] ESP32-C2
- [x] ESP32-C3
- [x] ESP32-C6
- [x] ESP32-H2
- [x] ESP32-S2
- [x] ESP32-S3

### PR DESCRIPTION
**Pull request by bugadani**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [ ] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description

The hardware has some difficulties with mode 0 and 2, so those aren't supported. According to the ESP-IDF [docs](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/spi_slave.html#restrictions-and-known-issues) we could make it work without DMA (and shovel FIFO contents into the buffer in interrupt handlers, for example).

#### Testing

Enabled the test case for ESP32, modified it to use Mode 1 and ran it locally (with a different pin assignment suitable for the Ethernet devkits I have).

Fixes #469

### CODE REVIEW

File: esp-hal/src/spi/slave.rs [hunk 1]
```diff
@@ -22,8 +22,8 @@
 //! # use esp_hal::dma::Dma;
 //! # use esp_hal::gpio::Io;
 //! let dma = Dma::new(peripherals.DMA);
-#![cfg_attr(esp32s2, doc = "let dma_channel = dma.spi2channel;")]
-#![cfg_attr(not(esp32s2), doc = "let dma_channel = dma.channel0;")]
+#![cfg_attr(pdma, doc = "let dma_channel = dma.spi2channel;")]
```
Comment (bjoernQ) [inline esp-hal/src/spi/slave.rs:25]: not(gdma) ?

Comment (bugadani) [reply esp-hal/src/spi/slave.rs:25]: Do you expect future devices to return to per-peripheral DMA channels?

Comment (bjoernQ) [reply esp-hal/src/spi/slave.rs:25]: ah no .... I read `gdma` for some reason here .... time for another coffee then

File: esp-hal/src/soc/esp32/gpio.rs [hunk 2]
```diff
@@ -231,29 +226,6 @@ pub enum InputSignal {
     PWM1_CAP0             = 112,
     PWM1_CAP1             = 113,
     PWM1_CAP2             = 114,
-    PWM2_FLTA             = 115,
```
Comment (MabezDev) [inline esp-hal/src/soc/esp32/gpio.rs:234]: Did you mean to remove these?

Comment (bugadani) [reply esp-hal/src/soc/esp32/gpio.rs:234]: yup, they aren't present in the TRM and they aren't used in the HAL.

**Review by MabezDev — APPROVED** [review]
LGTM, I think panicking in the case of passing an unsupported mode for esp32 is a good idea.

File: esp-hal/src/spi/slave.rs [hunk 3]
```diff
@@ -656,71 +633,106 @@ pub trait Instance: private::Sealed {
...
+                w.ck_idle_edge()
+                    .bit(matches!(data_mode, SpiMode::Mode0 | SpiMode::Mode1))
+            });
+            reg_block.user().modify(|_, w| {
+                w.ck_i_edge()
+                    .bit(matches!(data_mode, SpiMode::Mode1 | SpiMode::Mode2))
+            });
+            reg_block.ctrl2().modify(|_, w| unsafe {
+                match data_mode {
+                    SpiMode::Mode0 => {
```
Comment (JurajSadel) [inline esp-hal/src/spi/slave.rs:667]: Does it make sense to "configure" Mode0 and 2 here?

Comment (bugadani) [reply esp-hal/src/spi/slave.rs:667]: Yes, in case we add CPU-controlled mode we won't have to debug why ESP32 panics :)

**Review by JurajSadel — APPROVED** [review]
Thanks!

**Pull request merged**